File: tools/train_letter_writing_model.mjs
#!/usr/bin/env node
/**
 * Train shared letter-writing CNN from teacher bootstrap seed (+ optional judgments).
 *
 *   1. Export seed from app → data/writing-bank/teacher-seed.json
 *   2. node tools/train_letter_writing_model.mjs
 *
 * Writes public/models/letter-writing/ (manifest + TF.js). Architecture matches
 * src/letter-writing-tf.ts createModel().
 */
import * as tf from '@tensorflow/tfjs';
import fs from 'fs/promises';
import path from 'path';
import { fileURLToPath } from 'url';
import { RASTER_SIZE, rasterizeStrokesNode } from './letter_writing_raster.mjs';

const ROOT = path.join(path.dirname(fileURLToPath(import.meta.url)), '..');
const STAGE_ID = 'letter-writing';
const NUM_CLASSES = 26;
const LETTERS = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'.split('');
const BANK_PATH = path.join(ROOT, 'data', 'writing-bank', 'teacher-seed.json');
const CALIBRATION_DIR = path.join(ROOT, 'data', 'writing-calibration');
const STAGE_DIR = path.join(ROOT, 'public', 'models', STAGE_ID);

function letterToIndex(letter) {
  const code = String(letter).toUpperCase().charCodeAt(0);
  if (code < 65 || code > 90) return -1;
  return code - 65;
}

/** @returns {tf.LayersModel} — keep in sync with letter-writing-tf.ts createModel() */
function createModel() {
  const m = tf.sequential();
  m.add(
    tf.layers.conv2d({
      inputShape: [RASTER_SIZE, RASTER_SIZE, 1],
      filters: 16,
      kernelSize: 3,
      activation: 'relu',
      padding: 'same',
    }),
  );
  m.add(tf.layers.maxPooling2d({ poolSize: 2 }));
  m.add(tf.layers.conv2d({ filters: 32, kernelSize: 3, activation: 'relu', padding: 'same' }));
  m.add(tf.layers.maxPooling2d({ poolSize: 2 }));
  m.add(tf.layers.conv2d({ filters: 48, kernelSize: 3, activation: 'relu', padding: 'same' }));
  m.add(tf.layers.maxPooling2d({ poolSize: 2 }));
  m.add(tf.layers.flatten());
  m.add(tf.layers.dense({ units: 96, activation: 'relu' }));
  m.add(tf.layers.dropout({ rate: 0.3 }));
  m.add(tf.layers.dense({ units: NUM_CLASSES, activation: 'softmax' }));
  m.compile({
    optimizer: tf.train.adamax(0.001),
    loss: 'categoricalCrossentropy',
    metrics: ['accuracy'],
  });
  return m;
}

async function loadBootstrapSamples() {
  let raw;
  try {
    raw = await fs.readFile(BANK_PATH, 'utf8');
  } catch {
    throw new Error(
      `Missing ${BANK_PATH}\n` +
        'Export from the app: collector panel → Export writing seed, save as data/writing-bank/teacher-seed.json',
    );
  }
  const bank = JSON.parse(raw);
  const samples = bank.samples ?? bank;
  const xs = [];
  const ys = [];

  for (const letter of LETTERS) {
    const sets = samples[letter];
    if (!sets?.length) {
      throw new Error(`Bootstrap seed missing letter ${letter} in teacher-seed.json`);
    }
    const idx = letterToIndex(letter);
    for (const strokes of sets) {
      const flat = rasterizeStrokesNode(strokes);
      for (let r = 0; r < 4; r++) {
        xs.push(flat);
        ys.push(idx);
      }
    }
  }
  return { xs, ys };
}

async function loadJudgmentSamples(xs, ys, seen) {
  let files = [];
  try {
    files = await fs.readdir(CALIBRATION_DIR);
  } catch {
    return 0;
  }
  let added = 0;
  for (const name of files.filter((f) => f.endsWith('.json'))) {
    const row = JSON.parse(await fs.readFile(path.join(CALIBRATION_DIR, name), 'utf8'));
    if (row.teacherPass !== true && row.pass !== true) continue;
    const letter = row.letter ?? row.targetLetter;
    const idx = letterToIndex(letter);
    if (idx < 0 || !row.strokes?.length) continue;
    const flat = rasterizeStrokesNode(row.strokes);
    const key = `${idx}:${flat[0].toFixed(4)}:${flat.length}`;
    if (seen.has(key)) continue;
    seen.add(key);
    xs.push(flat);
    ys.push(idx);
    added++;
  }
  return added;
}

function normalizeManifestVersion(raw) {
  const v = Number(raw);
  if (!Number.isFinite(v)) return 0;
  if (v >= 1 && v < 10 && Number.isInteger(v)) return Math.round((v / 10) * 100) / 100;
  return Math.round(v * 100) / 100;
}

function bumpMinorVersion(current) {
  return Math.round(normalizeManifestVersion(current) * 100 + 1) / 100;
}

async function readManifestVersion() {
  try {
    const m = JSON.parse(await fs.readFile(path.join(STAGE_DIR, 'manifest.json'), 'utf8'));
    return normalizeManifestVersion(m.version ?? 0);
  } catch {
    return 0;
  }
}

async function exportTfjsModel(model) {
  await model.save(
    tf.io.withSaveHandler(async (artifacts) => {
      const weightData = artifacts.weightData;
      const buffer =
        weightData instanceof ArrayBuffer
          ? Buffer.from(weightData)
          : Buffer.concat(weightData.map((ab) => Buffer.from(ab)));

      const shard = 'group1-shard1of1.bin';
      const weightsManifest = [
        {
          paths: [shard],
          weights: artifacts.weightSpecs.map((w) => ({
            name: w.name,
            shape: w.shape,
            dtype: w.dtype,
          })),
        },
      ];

      const modelJson = {
        format: 'layers-model',
        generatedBy: 'EARLY tools/train_letter_writing_model.mjs',
        convertedBy: null,
        modelTopology: artifacts.modelTopology,
        weightsManifest,
      };

      await fs.mkdir(STAGE_DIR, { recursive: true });
      await fs.writeFile(path.join(STAGE_DIR, 'model.json'), JSON.stringify(modelJson));
      await fs.writeFile(path.join(STAGE_DIR, shard), buffer);

      return {
        modelArtifactsInfo: {
          dateSaved: new Date(),
          modelTopologyType: 'JSON',
        },
      };
    }),
  );
}

async function main() {
  await tf.setBackend('cpu');
  await tf.ready();

  const { xs, ys } = await loadBootstrapSamples();
  const seen = new Set();
  const judgmentCount = await loadJudgmentSamples(xs, ys, seen);

  const xStack = tf.stack(xs.map((f) => tf.tensor3d(f, [RASTER_SIZE, RASTER_SIZE, 1])));
  const yOneHot = tf.oneHot(tf.tensor1d(ys, 'int32'), NUM_CLASSES);

  const model = createModel();
  console.log(
    `Training letter-writing CNN: ${xs.length} rasters (${LETTERS.length} bootstrap letters, +${judgmentCount} judgments)`,
  );

  await model.fit(xStack, yOneHot, {
    epochs: 40,
    batchSize: Math.min(16, xs.length),
    shuffle: true,
    verbose: 1,
  });

  xStack.dispose();
  yOneHot.dispose();

  await exportTfjsModel(model);
  model.dispose();

  const nextVersion = bumpMinorVersion(await readManifestVersion());
  const manifest = {
    version: nextVersion,
    stageId: STAGE_ID,
    modelUrl: `/models/${STAGE_ID}/model.json`,
    trainedAt: new Date().toISOString(),
    sampleCount: xs.length,
    bootstrapLetters: LETTERS.length,
    judgmentSamples: judgmentCount,
  };
  await fs.writeFile(path.join(STAGE_DIR, 'manifest.json'), JSON.stringify(manifest, null, 2) + '\n');

  console.log(`Wrote public/models/${STAGE_ID}/ — v${nextVersion}, ${xs.length} training rasters.`);
  console.log('Commit public/models/letter-writing/ and deploy so all devices load the shared model.');
}

main().catch((err) => {
  console.error(err.message || err);
  process.exit(1);
});

File: tools/train_letter_writing_model.py
#!/usr/bin/env python3
"""
Train shared letter-writing CNN from teacher bootstrap seed (+ optional judgments).

  1. Export seed from app → data/writing-bank/teacher-seed.json
  2. python tools/train_letter_writing_model.py

Writes train-weights.json, then Node exports TF.js (tools/export_letter_writing_tfjs.mjs).

Requires: pip install -r tools/requirements-train.txt
"""

from __future__ import annotations

import hashlib
import json
import subprocess
import sys
from datetime import datetime, timezone
from pathlib import Path

import numpy as np
import tensorflow as tf

from letter_writing_raster import RASTER_SIZE, letter_to_index, rasterize_strokes
from model_version import bump_minor_version, normalize_manifest_version

ROOT = Path(__file__).resolve().parents[1]
STAGE_ID = "letter-writing"
LETTERS = list("ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz")
NUM_CLASSES = 52
BANK_PATH = ROOT / "data" / "writing-bank" / "teacher-seed.json"
CALIBRATION_DIR = ROOT / "data" / "writing-calibration"
CALIBRATION_ARCHIVE_DIR = ROOT / "data" / "training-archive" / "writing-calibration"
STAGE_DIR = ROOT / "public" / "models" / STAGE_ID
# Mirror bootstrap: each accepted judgment is repeated this many times so
# teacher corrections are not drowned out by 4× seed copies per letter.
JUDGMENT_COPIES = 4


def build_model() -> tf.keras.Model:
    """Keep in sync with letter-writing-tf.ts createModel()."""
    return tf.keras.Sequential(
        [
            tf.keras.layers.Conv2D(
                16, 3, activation="relu", padding="same", input_shape=(RASTER_SIZE, RASTER_SIZE, 1)
            ),
            tf.keras.layers.MaxPooling2D(2),
            tf.keras.layers.Conv2D(32, 3, activation="relu", padding="same"),
            tf.keras.layers.MaxPooling2D(2),
            tf.keras.layers.Conv2D(48, 3, activation="relu", padding="same"),
            tf.keras.layers.MaxPooling2D(2),
            tf.keras.layers.Flatten(),
            tf.keras.layers.Dense(96, activation="relu"),
            tf.keras.layers.Dropout(0.3),
            tf.keras.layers.Dense(NUM_CLASSES, activation="softmax"),
        ],
        name="letter_writing_cnn",
    )


def load_bootstrap_samples() -> tuple[list[list[float]], list[int]]:
    if not BANK_PATH.is_file():
        raise SystemExit(
            f"Missing {BANK_PATH}\n"
            "Export from the app: collector panel → Export writing seed, "
            "save as data/writing-bank/teacher-seed.json"
        )

    bank = json.loads(BANK_PATH.read_text(encoding="utf-8"))
    samples: dict[str, list] = bank.get("samples", bank)
    xs: list[list[float]] = []
    ys: list[int] = []

    for letter in LETTERS:
        sets = samples.get(letter)
        if not sets:
            raise SystemExit(f"Bootstrap seed missing letter {letter} in teacher-seed.json")
        idx = letter_to_index(letter)
        for strokes in sets:
            flat = rasterize_strokes(strokes)
            for _ in range(4):
                xs.append(flat)
                ys.append(idx)

    return xs, ys


def _iter_calibration_rows(path: Path):
    """Yield individual sample dicts from either a single-object or array/payload file."""
    raw = json.loads(path.read_text(encoding="utf-8"))
    if isinstance(raw, list):
        for item in raw:
            if isinstance(item, dict):
                yield item
    elif isinstance(raw, dict):
        samples = raw.get("samples")
        if isinstance(samples, list):
            for item in samples:
                if isinstance(item, dict):
                    yield item
        else:
            yield raw


def _judgment_dedupe_key(row: dict, flat: list[float], class_idx: int) -> str:
    """
    Unique key per teacher-accepted attempt.

    Do NOT use flat[0] — with canvas-relative rasters the top-left pixel is
    almost always background (0), so every judgment for "o" looked identical
    and only one survived training.
    """
    attempt_id = row.get("attemptId")
    if attempt_id:
        return f"id:{attempt_id}"
    digest = hashlib.sha256(np.array(flat, dtype=np.float32).tobytes()).hexdigest()[:16]
    return f"r:{class_idx}:{digest}"


def load_judgment_samples(
    xs: list[list[float]], ys: list[int], seen: set[str]
) -> tuple[int, int, int]:
    """
    Load extra training samples from data/writing-calibration/*.json.

    Only attempts where the teacher explicitly accepted (teacherPass === true)
    are used. Model self-accepts and heuristic passes are NEVER consumed here —
    that would let the model reinforce its own mistakes.

    Returns (raster_count_added, judgment_files_read, duplicate_attempts_skipped).
    """
    judgment_dirs = [d for d in (CALIBRATION_ARCHIVE_DIR, CALIBRATION_DIR) if d.is_dir()]
    if not judgment_dirs:
        return 0, 0, 0

    files_read = 0
    duplicate_attempts = 0
    unique_judgments = 0
    rasters_added = 0

    paths: list[Path] = []
    for judgment_dir in judgment_dirs:
        paths.extend(sorted(judgment_dir.glob("*.json")))

    for path in paths:
        try:
            rows = list(_iter_calibration_rows(path))
        except json.JSONDecodeError as err:
            print(f"  skipping unreadable calibration file {path.name}: {err}")
            continue
        files_read += 1
        for row in rows:
            if row.get("teacherPass") is not True:
                continue
            letter = row.get("letter") or row.get("targetLetter")
            idx = letter_to_index(str(letter or ""))
            strokes = row.get("strokes") or []
            if idx < 0 or not strokes:
                continue
            flat = rasterize_strokes(strokes)
            key = _judgment_dedupe_key(row, flat, idx)
            if key in seen:
                duplicate_attempts += 1
                continue
            seen.add(key)
            unique_judgments += 1
            for _ in range(JUDGMENT_COPIES):
                xs.append(flat)
                ys.append(idx)
                rasters_added += 1

    if files_read:
        print(
            f"  Judgments: {files_read} files -> {unique_judgments} unique accepts "
            f"({duplicate_attempts} true duplicates skipped) "
            f"-> {rasters_added} training rasters (x{JUDGMENT_COPIES})",
            flush=True,
        )

    return rasters_added, files_read, duplicate_attempts


def read_manifest_version() -> float:
    manifest_path = STAGE_DIR / "manifest.json"
    if not manifest_path.is_file():
        return 0.0
    try:
        data = json.loads(manifest_path.read_text(encoding="utf-8"))
        return normalize_manifest_version(float(data.get("version", 0)))
    except (json.JSONDecodeError, TypeError, ValueError):
        return 0.0


def export_train_weights(model: tf.keras.Model, path: Path) -> None:
    layers: list[dict] = []
    for layer in model.layers:
        weights = layer.get_weights()
        if not weights:
            continue
        name = layer.__class__.__name__
        if name == "Conv2D":
            kernel, bias = weights
            layers.append(
                {
                    "type": "conv2d",
                    "kernel": kernel.tolist(),
                    "bias": bias.ravel().tolist(),
                }
            )
        elif name == "Dense":
            kernel, bias = weights
            layers.append(
                {
                    "type": "dense",
                    "kernel": kernel.tolist(),
                    "bias": bias.ravel().tolist(),
                }
            )
        else:
            raise SystemExit(f"Unexpected trainable layer type: {name}")

    if len(layers) != 5:
        raise SystemExit(f"Expected 5 weight layers (3 conv + 2 dense), got {len(layers)}")

    spec = {
        "rasterSize": RASTER_SIZE,
        "numClasses": NUM_CLASSES,
        "layers": layers,
    }
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(json.dumps(spec, indent=2) + "\n", encoding="utf-8")


def main() -> None:
    xs, ys = load_bootstrap_samples()
    bootstrap_rasters = len(xs)
    seen: set[str] = set()
    judgment_rasters, judgment_files, _judgment_dupes = load_judgment_samples(xs, ys, seen)
    unique_judgments = judgment_rasters // JUDGMENT_COPIES if judgment_rasters else 0

    x_arr = np.array(xs, dtype=np.float32).reshape(-1, RASTER_SIZE, RASTER_SIZE, 1)
    y_arr = tf.keras.utils.to_categorical(ys, NUM_CLASSES)

    print(
        f"Training letter-writing CNN: {len(xs)} rasters "
        f"({bootstrap_rasters} from seed, +{judgment_rasters} from "
        f"{unique_judgments} judgments in {judgment_files} files)",
        flush=True,
    )

    model = build_model()
    model.compile(
        optimizer=tf.keras.optimizers.Adamax(learning_rate=0.001),
        loss="categorical_crossentropy",
        metrics=["accuracy"],
    )
    model.fit(
        x_arr,
        y_arr,
        epochs=40,
        batch_size=min(16, len(xs)),
        shuffle=True,
        verbose=1,
    )

    weights_path = STAGE_DIR / "train-weights.json"
    export_train_weights(model, weights_path)

    export_script = ROOT / "tools" / "export_letter_writing_tfjs.mjs"
    subprocess.run(["node", str(export_script)], cwd=ROOT, check=True)
    weights_path.unlink(missing_ok=True)

    next_version = bump_minor_version(read_manifest_version())
    manifest = {
        "version": next_version,
        "stageId": STAGE_ID,
        "modelUrl": f"/models/{STAGE_ID}/model.json",
        "trainedAt": datetime.now(timezone.utc).isoformat(),
        "sampleCount": len(xs),
        "bootstrapLetters": len(LETTERS),
        "judgmentSamples": unique_judgments,
        "judgmentRasters": judgment_rasters,
    }
    (STAGE_DIR / "manifest.json").write_text(json.dumps(manifest, indent=2) + "\n", encoding="utf-8")

    print(f"Wrote public/models/{STAGE_ID}/ — v{next_version}, {len(xs)} training rasters.")
    print("Commit public/models/letter-writing/ and deploy so all devices load the shared model.")


if __name__ == "__main__":
    try:
        main()
    except KeyboardInterrupt:
        sys.exit(130)
